tests/sleepq_abort.c

```diff
@@ -0,0 +1,118 @@
+#include <sleepq.h>
+#include <runq.h>
+#include <ktest.h>
+#include <sched.h>
+
+#define T 6
+
+#define RAND_COUNT 13
+/* obtained with totally fair d6 rolls (1-3 true; 4-6 false) */
+static bool kinda_random_values[RAND_COUNT] = {true,  true,  false, true, false,
+                                               false, true,  false, true, true,
+                                               false, false, false};
+/* some random ordering for waiters (values from 0 to T-1) */
+static int waiters_ord[T] = {5, 2, 1, 3, 4, 0};
+static thread_t *waiters[T];
+static thread_t *waker;
+/* just to have some unused waiting channel */
+static int some_val;
+static volatile int woken_gracefully;
+static volatile int interrupted;
+
+/* Waiters have higher priority so that waker will only execute
+ * when waiters can't.
+ * Therefore there should be only one waiter active at once */
+static void waiter_routine(void *_arg) {
+  slp_wakeup_t rsn = sleepq_wait_abortable(&some_val, __caller(0), SLPF_INT);
+
+  if (rsn == SLEEPQ_WKP_INT)
+    interrupted++;
+  else if (rsn == SLEEPQ_WKP_REG)
+    woken_gracefully++;
+  else
+    assert(false);
+}
+
+static void waker_routine(void *_arg) {
+  int woken = 0; /* total */
+  int aborted = 0;
+  int next_abort = 0;
+  int rand_next = 0;
+
+  while (woken < T) {
+    /* this is a terrible way for randomizing variables
+     * (especially when it's probably not even needed) */
+    bool wake = kinda_random_values[rand_next % RAND_COUNT];
+    rand_next++;
+
+    if (wake) {
+      bool succ = sleepq_signal(&some_val);
+      assert(succ);
+      aborted++;
+      woken++;
+    } else {
+      bool succ = false;
+      while (!succ) {
+        assert(next_abort < T && waiters_ord[next_abort] < T);
+        succ = sleepq_abort(waiters[waiters_ord[next_abort]], SLEEPQ_WKP_INT);
+        next_abort++;
+      }
+      woken++;
+    }
```

Comment:
Maybe increase `woken` once, after `if-else`?

Reply:
I think I prefer having it twice for explicitness.